File: java/kmm/parser/SomeParsers.java
package kmm.parser;

public class SomeParsers {
    public final static Parser<Character> alnum = Parser.satisfy(c -> Character.isLetterOrDigit(c) || c == '_');

    public static Parser<Character> char_(char c){
        return Parser.satisfy(x -> x == c);
    }

    public static Parser<Character> symbol(char c){
        return char_(c).token();
    }

    public static Parser<String> name(String n){
        return alnum.some().flatMap(s -> s.equals(n) ? Parser.pure(n) : Parser.empty()).token();
    }

    public static Parser<String> optional_s(Parser<String> p){
        return p.orElse(Parser.pure(""));
    }

    public static Parser<String> optional_c(Parser<Character> p){
        return optional_s(p.map(Object::toString));
    }

    public final static Parser<Character> digit = Parser.satisfy(Character::isDigit);

    public final static Parser<String> digits = digit.many();

    public final static Parser<String> sign = optional_c(char_('+').orElse(char_('-')));

    public final static Parser<String> usign = optional_c(symbol('+').orElse(symbol('-')));

    public final static Parser<Double> double_ = sign.flatMap(
        sign_part -> digits.flatMap(
        int_part  -> optional_s(char_('.').skip(digits)).flatMap(
        frac_part -> optional_s(char_('e').orElse(char_('E')).skip(sign).flatMap(
            exp_sign -> digit.some().flatMap(
            exp_digits -> Parser.pure(exp_sign + exp_digits)))).flatMap(
        exp_part -> !int_part.isEmpty() || !frac_part.isEmpty() ?
            Parser.pure(Double.valueOf(sign_part + int_part +
                (!frac_part.isEmpty() ? '.' + frac_part : "") +
                (!exp_part.isEmpty() ? 'e' + exp_part : ""))) :
            Parser.empty())))).token();

    public final static Parser<Double> _double_ = Parser.satisfy(Character::isDigit).some()
        .flatMap(s -> Parser.pure(Double.valueOf(s))).token();
}

File: F#/SomeParsers.fs
module SomeParsers

open System // Char
open Parser

let alnum: Parser<char> = Parser<char>.Satisfy(fun c -> Char.IsLetterOrDigit(c) || c.Equals '_')

let Name (n: string): Parser<string> = alnum.Some().FlatMap(fun s -> if s.Equals(n) then Parser.Pure(n) else Parser<string>.Empty()).Token()

let Char_(c: char): Parser<char> = Parser<char>.Satisfy(_.Equals(c))

let Symbol(c: char): Parser<char> = Char_(c).Token()

let digits: Parser<string> = (Parser<char>.Satisfy Char.IsDigit).Many()

let sign: Parser<string> = Parser<char>.Optional_c(Char_('+').OrElse(Char_ '-'))
let usign: Parser<string> = Parser<char>.Optional_c(Symbol('+').OrElse(Symbol '-'))

let NumberDecimalSeparator = System.Globalization.CultureInfo.CurrentCulture.NumberFormat.NumberDecimalSeparator

let double: Parser<float> = digits.FlatMap(
    fun int_part  -> Parser<string>.Optional_s(Char_('.').Skip(digits)).FlatMap(
    fun frac_part -> Parser<string>.Optional_s(Char_('e').OrElse(Char_ 'E').Skip(sign).FlatMap(
        fun exp_sign   -> Parser<string>.Satisfy(Char.IsDigit).Some().FlatMap(
        fun exp_digits -> Parser<string>.Pure(exp_sign + exp_digits)))).FlatMap(
    fun exp_part ->
        if int_part.Length > 0 || frac_part.Length > 0
        then Parser<float>.Pure(Double.Parse(int_part +
            (if frac_part.Length > 0 then NumberDecimalSeparator + frac_part else "") +
            (if exp_part.Length > 0 then "e" + exp_part else "")))
        else Parser<float>.Empty()))).Token()
